browser/brave_rewards/checkout_dialog.cc

```diff
@@ -0,0 +1,298 @@
+    case DialogCloseReason::UnverifiedWallet:
+      request_->TerminateConnectionWithMessage(
+          PaymentErrorReason::NOT_SUPPORTED, errors::kUnverifiedUserWallet);
+      break;
+    case DialogCloseReason::InsufficientBalance:
+      request_->TerminateConnectionWithMessage(
+          PaymentErrorReason::NOT_SUPPORTED, errors::kInsufficientBalance);
+      break;
+  }
+}
+
+void CheckoutDialogDelegate::OnCloseContents(WebContents* source,
+                                             bool* out_close_dialog) {
+  *out_close_dialog = true;
+}
+
+bool CheckoutDialogDelegate::ShouldShowDialogTitle() const {
+  return false;
+}
+
+CheckoutDialogHandler::CheckoutDialogHandler(
+    base::WeakPtr<PaymentRequest> request)
+    : request_(request),
+      weak_factory_(this) {}
+
+CheckoutDialogHandler::~CheckoutDialogHandler() = default;
+
+RewardsService* CheckoutDialogHandler::GetRewardsService() {
+  if (!rewards_service_) {
+    Profile* profile = Profile::FromWebUI(web_ui());
+    rewards_service_ = RewardsServiceFactory::GetForProfile(profile);
+    rewards_service_->StartProcess(base::DoNothing());
+  }
+  return rewards_service_;
+}
+
+void CheckoutDialogHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "paymentRequestComplete",
+      base::BindRepeating(&CheckoutDialogHandler::HandlePaymentCompletion,
+                          base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "getWalletBalance",
+      base::BindRepeating(&CheckoutDialogHandler::OnGetWalletBalance,
+                          base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "getExternalWallet",
+      base::BindRepeating(&CheckoutDialogHandler::GetExternalWallet,
+                          base::Unretained(this)));
+  web_ui()->RegisterMessageCallback(
+      "getRewardsParameters",
+      base::BindRepeating(&CheckoutDialogHandler::GetRewardsParameters,
+                          base::Unretained(this)));
+}
+
+void CheckoutDialogHandler::HandlePaymentCompletion(
+    const base::ListValue* args) {
+  auto spec = request_->spec();
+  if (!request_->spec()->IsInitialized()) {
+    return;
+  }
+
+  auto* rfh =
+      content::RenderFrameHost::FromID(request_->initiator_frame_routing_id());
+  if (!rfh) {
+    request_->TerminateConnectionWithMessage(
+        PaymentErrorReason::INVALID_DATA_FROM_RENDERER,
+        errors::kInvalidRenderer);
+    return;
+  }
+
+  auto* rewards_service = brave_rewards::RewardsServiceFactory::GetForProfile(
+      Profile::FromBrowserContext(rfh->GetBrowserContext()));
+  if (!rewards_service) {
+    request_->TerminateConnectionWithMessage(
+        PaymentErrorReason::INVALID_DATA_FROM_RENDERER,
+        errors::kRewardsNotInitialized);
+    return;
+  }
+
+  const auto& display_items =
+      spec->GetDisplayItems(request_->state()->selected_app());
+  for (size_t i = 0; i < display_items.size(); i++) {
+    DCHECK((*display_items[i])->sku.has_value());
```

Comment:
why is this a DCHECK? An empty sku would be invalid, right? I can't remember what happens if you call `value()` when `has_value() == false`, but it seems like a problem and we should be aborting here, not adding the empty value

Reply:
Code here checks if the items have valid `sku` tokens: https://github.com/brave/brave-core/pull/7557/files#diff-12d68f7aee6cfd7305f59bcf92a37b1fa2353b1bd0a8b5224c7f4b63c68bfd2bR47 - `ShowCheckoutDialog` should only be invoked if `displayItems` have valid `sku` tokens.